src/Storage/Controllers/CleanupController.cs

```diff
@@ -0,0 +1,246 @@
+    [ApiController]
+    public class CleanupController : ControllerBase
+    {
+        private readonly IInstanceRepository _instanceRepository;
+        private readonly IApplicationRepository _applicationRepository;
+        private readonly IBlobRepository _blobRepository;
+        private readonly IDataRepository _dataRepository;
+        private readonly IInstanceEventRepository _instanceEventRepository;
+        private readonly ILogger<CleanupController> _logger;
+        private readonly bool _usePostgreSQL;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CleanupController"/> class
+        /// </summary>
+        /// <param name="instanceRepository">the instance repository handler</param>
+        /// <param name="applicationRepository">the application repository handler</param>
+        /// <param name="blobRepository">the blob repository handler</param>
+        /// <param name="dataRepository">the data repository handler</param>
+        /// <param name="instanceEventRepository">the instance event repository handler</param>
+        /// <param name="logger">the logger</param>
+        /// <param name="generalSettings">the general settings.</param>
+        public CleanupController(
+            IInstanceRepository instanceRepository,
+            IApplicationRepository applicationRepository,
+            IBlobRepository blobRepository,
+            IDataRepository dataRepository,
+            IInstanceEventRepository instanceEventRepository,
+            ILogger<CleanupController> logger,
+            IOptions<GeneralSettings> generalSettings)
+        {
+            _instanceRepository = instanceRepository;
+            _applicationRepository = applicationRepository;
+            _blobRepository = blobRepository;
+            _dataRepository = dataRepository;
+            _instanceEventRepository = instanceEventRepository;
+            _logger = logger;
+            _usePostgreSQL = generalSettings.Value.UsePostgreSQL;
+        }
+
+        /// <summary>
+        /// Invoke periodic cleanup of instances
+        /// </summary>
+        /// <returns>?</returns>
+        [HttpDelete("cleanupinstances")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ApiExplorerSettings(IgnoreApi = true)]
+        public async Task<ActionResult> CleanupInstances()
+        {
+            if (!_usePostgreSQL)
+            {
+                return Ok();
+            }
+
+            List<Instance> instances = await _instanceRepository.GetHardDeletedInstances();
+            List<string> autoDeleteAppIds = (await _applicationRepository.FindAll())
+                .Where(a => instances.Select(i => i.AppId).ToList().Contains(a.Id) && a.AutoDeleteOnProcessEnd)
+                .Select(a => a.Id).ToList();
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            int successfullyDeleted = await CleanupInstancesInternal(instances, autoDeleteAppIds, "NightlyCleanup");
```

Comment:
We're inconsistent in how we handle the "caller" parameter. Here it is hard coded.